download/download.go

```diff
@@ -0,0 +1,163 @@
+	runId := context.Arguments["<runId>"].(string)
+	artifact := context.Arguments["<artifact>"].(string)
+
+	provider := extpoints.CommandProviders()[command]
+	if provider == nil {
+		log.Panicf("Unknown command %s", command)
+
+	} else {
+
+		permaCred := &tcclient.Credentials{
+			ClientID:    os.Getenv("TASKCLUSTER_CLIENT_ID"),
+			AccessToken: os.Getenv("TASKCLUSTER_ACCESS_TOKEN"),
+		}
+
+		userQueue := queue.New(permaCred)
+
+		if runId != "" {
+			//get a artifact with runId parameter
+			url_artifact, err := userQueue.GetArtifact_SignedURL(taskId, runId, artifact, time.Second*10)
+			if err != nil {
+				log.Panicf("Exception thrown signing URL \n%s", err)
+			} else {
+
+				url_artifact := EnforceHttpsUrl(url_artifact.String())
+				fmt.Printf("NEW URL %s\n", url_artifact)
+				response, attempts, err := getAnArtifact(url_artifact)
+
+				if err != nil {
+					log.Panicf("Exception thrown download an artifact \n%s", err)
+				} else {
+					fmt.Printf("Number of attempts: %d\n", attempts)
+					_, length, out := checkContentLength(response)
+					log.Printf("ContentLength %d with %s", length, out)
+				}
+			}
+		}
+		if runId == "" {
+			//get latest artifact without rundId parameter
+			url_artifact, err := userQueue.GetLatestArtifact_SignedURL(taskId, artifact, time.Second*10)
+			if err != nil {
+				log.Panicf("Exception thrown signing URL \n%s", err)
+			} else {
+
+				url_artifact := EnforceHttpsUrl(url_artifact.String())
+				//fmt.Printf("NEW URL %s\n", url_artifact)
+				response, attempts, err := getAnArtifact(url_artifact)
+
+				if err != nil {
+					log.Panicf("Exception thrown download an artifact \n%s", err)
+				} else {
+					fmt.Printf("Number of attempts: %d\n", attempts)
+					fmt.Printf("Content:::: \n %s \n", response.Header.Get("Content-Type"))
+					_, length, out := checkContentLength(response)
+					log.Printf("ContentLength %d with %s\n", length, out)
+				}
+			}
+		}
+
+	}
+	return true
+}
+
+func getAnArtifact(url string) (*http.Response, int, error) {
+	res, attempts, err := httpbackoff.Retry(func() (*http.Response, error, error) {
+		resp, err := http.Get(url)
+		// assume all errors are temporary
+
+		//following redirect if there is a new url, link, redirect
+		return resp, err, nil
+	})
+	return res, attempts, err
+}
+
+func checkContentLength(res *http.Response) (error, int64, string) {
+
+	if res.ContentLength > 0 {
+		return nil, res.ContentLength, "Good"
+	}
+	if res.ContentLength == 0 {
+		//Means exactly none
+		if res.Body != nil {
+			return nil, res.ContentLength, "None With Some Body Content"
+		}
+		return nil, res.ContentLength, "None"
+	}
+	if res.ContentLength < 0 {
+		//Means Unknown
+		return nil, res.ContentLength, "Chunked"
+	}
+	return nil, 0, ""
+}
+
+func StreamingArtifactFile(res *http.Response) bool {
+
+	return false
+
+}
+
+//function to enforce https if an artifact came without https assigned
+func EnforceHttpsUrl(url string) string {
```

Comment:
I am not sure if this is necessary, you may safely assume artifacts are https URLs. @jonasfj ?